File: Urdhva Triyagbhyam Sutra.py
import time


class Solution:
    def makeLengthEqual(self, a, b):
        if len(a) == len(b):
            return
        if len(a) < len(b):
            self.makeLengthEqual(b, a)
            return
        while len(b) != len(a):
            b = "0" + b

    def removeLeadingZero(self, num):
        n = len(num) - 1
        for i in range(len(num) - 1):
            if num[i] != '0':
                n = i
                break
        num = num[n:]

    def multiply(self, num1, num2):
        self.makeLengthEqual(num1, num2)
        n = len(num1)
        carry = 0
        ans = ""

        for i in range(n - 1, -1, -1):
            _sum = 0
            k = i
            for j in range(n - 1, i - 1, -1):
                _sum += int(num1[j]) * int(num2[k])
                k += 1
            _sum += carry
            carry = _sum // 10
            ans += str(_sum % 10)

        for i in range(n - 2, -1, -1):
            _sum = 0
            k = 0
            for j in range(i, -1, -1):
                _sum += int(num1[j]) * int(num2[k])
                k += 1
            _sum += carry
            carry = _sum // 10
            ans += str(_sum % 10)

        if carry:
            ans += str(carry)

        ans = ans[::-1]
        self.removeLeadingZero(ans)
        return ans


if __name__ == "__main__":
    solution = Solution()

    num1 = "114701722186227725530857903247788731566117812708912875331083418612132595800326"
    num2 = "987643221443489764384795643298746382985736482736487365847365874387458437564327"

    result = solution.multiply(num1, num2)

    print(f"The product of {num1} and {num2} is: {result}")


start_time = time.time()
# Run your algorithm here
end_time = time.time()

execution_time = end_time - start_time
print(f"Execution Time: {execution_time} seconds")

File: Urdhva Triyagbhyam Sutra.c
#include <stdio.h>
#include <string.h>
#include <time.h>


// Function to remove leading zeros from a string representing a number
void removeLeadingZero(char num[]) {
    int n = strlen(num) - 1;
    int i;
    for (i = 0; i < strlen(num) - 1; ++i) {
        if (num[i] != '0') {
            n = i;
            break;
        }
    }

    // Shifting the non-zero part to the beginning
    memmove(num, num + n, strlen(num) - n);
    num[strlen(num) - n] = '\0';
}

// Function to multiply two large numbers represented as strings
char* multiply(char num1[], char num2[], char result[]) {
    int len1 = strlen(num1);
    int len2 = strlen(num2);

    // Initializing the result array with zeros
    for (int i = 0; i < len1 + len2; ++i) {
        result[i] = '0';
    }
    result[len1 + len2] = '\0';

    // Multiplying each digit and add to the result
    for (int i = len1 - 1; i >= 0; --i) {
        int carry = 0;
        for (int j = len2 - 1; j >= 0; --j) {
            int temp = (num1[i] - '0') * (num2[j] - '0') + (result[i + j + 1] - '0') + carry;
            result[i + j + 1] = temp % 10 + '0';
            carry = temp / 10;
        }
        result[i] += carry; // Add the carry to the current digit
    }

    // Removing leading zeros from the result
    removeLeadingZero(result);

    return result;
}

int main() {
    char num1[] = "114701722186227725530857903247788731566117812708912875331083418612132595800326";
    char num2[] = "987643221443489764384795643298746382985736482736487365847365874387458437564327";
    char result[1000];

    clock_t start_time = clock(); // Record the start time

    multiply(num1, num2, result);

    clock_t end_time = clock(); // Record the end time
    
    multiply(num1, num2, result);

    printf("The product of %s and %s is: %s\n", num1, num2, result);
    
    double execution_time = ((double)(end_time - start_time)) / CLOCKS_PER_SEC;
    printf("Execution Time: %f seconds\n", execution_time);

    return 0;
}
